src/Features/Core/Portable/Workspace/ProjectCacheService.cs

```diff
@@ -55,6 +59,12 @@ public bool IsImplicitCacheEmpty
             }
         }
 
+        public void Dispose()
+        {
+            // cancel any pending blocks
+            _shutdownNotificationSource.Cancel();
```

Comment:
IThreadingContext.DisposalToken is tied to the life of the MEF catalog. This is potentially much longer than the life of the workspace. Items exported directly via MEF already have support for `IDisposable` during MEF catalog disposal, but services created _via factory_ for the context of a single workspace do not have the same support.

Reply:
> I'm also worried a bit here in that if the ImplicitCacheMonitor doesn't observe the cancellation token can it still run later, given that observing cancellation is usually considered a 'best effort'.

This operates very similar to [`WorkCoordinator.Shutdown`](https://github.com/dotnet/roslyn/blob/40adddcf0cac72eaa32170ac07817c6c42c55841/src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.cs#L123-L124), which is tied into [`Workspace.Dispose`](https://github.com/dotnet/roslyn/blob/40adddcf0cac72eaa32170ac07817c6c42c55841/src/Workspaces/Core/Portable/Workspace/Workspace.cs#L373). Unfortunately, I can't use `IEventListenerStoppable` for `ImplicitCacheMonitor` because it's [explicitly disabled](https://github.com/dotnet/roslyn/blob/40adddcf0cac72eaa32170ac07817c6c42c55841/src/Workspaces/CoreTestUtilities/Fakes/MockWorkspaceEventListenerProvider.cs#L12-L14) for most test scenarios. Switching to proper Workspace-owned `IDisposable` service implementations is much cleaner than trying to special case a new listener here.

Reply:
> IThreadingContext.DisposalToken is tied to the life of the MEF catalog. This is potentially much longer than the life of the workspace. 

My concern here is generally that we hae like 3 differnet types of "stop working" notifications one can be listening to.  Tehre's the mef catalog going down.  THis PR seems to be about the workspace going down.  And Tomas added support for "VS is shutting down".

It feels to me that we probably don't need all of these.  Don't these all boil down to "we're shutting down, stop working" and could all be exposed from a single source?

Reply:
Another way of putting this.  If you tied this to IThreadingContext.DisposalToken, what woudl be the issue?

Reply:
> If you tied this to IThreadingContext.DisposalToken, what woudl be the issue?

@CyrusNajmabadi See https://github.com/dotnet/roslyn/pull/47902#discussion_r492401218. During the life of the MEF catalog, an arbitrary number of workspaces can be created and disposed.

The more acute issue is `DisposalToken` is set _after_ pending operations complete (in an expedited manner) during tests. Tests are expected to dispose of workspace instances before the end of the test _and_ disposal of the workspace is expected to cancel operations that are no longer required. `DisposalToken` is triggered far after these events and represents a completely unrelated concept.

Reply:
Another way to look at this is `MefWorkspaceServices` supporting `IDisposable` is a natural extension of two things:

* The MEF `ExportProvider` supports `IDisposable`
* `MefWorkspaceServices` supports `IWorkspaceServiceFactory`

This functionality really should have been implemented from the very start, and naturally leads to a dramatic simplification of concepts like `IEventListenerStoppable` (though these simplifications will be part of separate pull requests sent to the master branch instead of here).

Reply:
So I understand @sharwell's argument about the lifetime differences, if we ignore the test situation for a moment then the common scenario would be the preview workspace spinning up and being destroyed repeatedly when you look at code fixes; those shouldn't live further.

I guess I had forgotten about IEventListenerStoppable though so now it seems we have two mechanisms here and I can't really reason about why we have two, and this:

> Unfortunately, I can't use IEventListenerStoppable for ImplicitCacheMonitor because it's explicitly disabled for most test scenarios.

seems like really we should ask ourselves why our unit tests are behaving different than the rest of the system in that way. Do we think IEventListenerStoppable is removable if we continue down this approach?

There's still a different bug I worry about here though. The fact our tests are blocking on async listeners between runs isn't something that we do in the actual product. If a Workspace instance in actual Visual Studio isn't disposed, can it be running and queuing tasks during VS shutdown and cause problems that way?

Reply:
> Do we think IEventListenerStoppable is removable if we continue down this approach?

Yes, it appears to be removable and I'm working to get that submitted as a follow-up change to master.